database/animals.ts

```diff
@@ -3,6 +3,64 @@ import { postgresToGraphql } from '../graphql/transform';
 import type { Animal } from '../migrations/00000-createTableAnimals';
 import { sql } from './connect';
 
+export const createAnimal = cache(
+  async (sessionToken: string, newAnimal: Omit<Animal, 'id'>) => {
+    const [animal] = await sql<Animal[]>`
+      INSERT INTO
+        animals (first_name, type, accessory) (
+          SELECT
+            ${newAnimal.firstName},
+            ${newAnimal.type},
+            ${newAnimal.accessory}
+          FROM
+            sessions
+          WHERE
+            token = ${sessionToken}
+            AND sessions.expiry_timestamp > now()
+        )
+      RETURNING
+        animals.*
+    `;
+
+    return postgresToGraphql(animal);
+  },
+);
+
+export const updateAnimal = cache(
+  async (sessionToken: string, updatedAnimal: Animal) => {
+    const [animal] = await sql<Animal[]>`
+      UPDATE animals
+      SET
+        first_name = ${updatedAnimal.firstName},
+        type = ${updatedAnimal.type},
+        accessory = ${updatedAnimal.accessory}
+      FROM
+        sessions
+      WHERE
+        sessions.token = ${sessionToken}
+        AND sessions.expiry_timestamp > now()
+        AND animals.id = ${updatedAnimal.id}
+      RETURNING
+        animals.*
+    `;
+    return postgresToGraphql(animal);
+  },
+);
+
+export const deleteAnimal = cache(async (sessionToken: string, id: number) => {
+  const [animal] = await sql<Animal[]>`
+    DELETE FROM animals USING sessions
+    WHERE
+      sessions.token = ${sessionToken}
+      AND sessions.expiry_timestamp > now()
+      AND animals.id = ${id}
+    RETURNING
+      animals.*
+  `;
+
+  return postgresToGraphql(animal);
+});
+
 export const getAnimalsInsecure = cache(async () => {
```

Comment:
This should probably have a comment with our usual pattern of noting the boundary of the insecure database query functions:

https://github.com/upleveled/next-js-example-fall-2024-atvie/blob/6a3b5fc5b95b6a2348e74d46d899c19273810c2c/database/animals.ts#L140-L143

(for all files)

Reply:
f3696c0ada47d4c5455090683a23a9d1dd4cfcdf